table/tables/partition.go

```diff
@@ -207,44 +212,53 @@ func generateHashPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo,
 	columns []*expression.Column, names types.NameSlice) (*PartitionExpr, error) {
 	var column *expression.Column
 	// The caller should assure partition info is not nil.
-	partitionPruneExprs := make([]expression.Expression, 0, len(pi.Definitions))
-	var buf bytes.Buffer
 	schema := expression.NewSchema(columns...)
-	for i := 0; i < int(pi.Num); i++ {
-		fmt.Fprintf(&buf, "MOD(ABS(%s),(%d))=%d", pi.Expr, pi.Num, i)
-		exprs, err := expression.ParseSimpleExprsWithNames(ctx, buf.String(), schema, names)
-		if err != nil {
-			// If it got an error here, ddl may hang forever, so this error log is important.
-			logutil.BgLogger().Error("wrong table partition expression", zap.String("expression", buf.String()), zap.Error(err))
-			return nil, errors.Trace(err)
-		}
-		// Get a hash code in advance to prevent data race afterwards.
-		exprs[0].HashCode(ctx.GetSessionVars().StmtCtx)
-		partitionPruneExprs = append(partitionPruneExprs, exprs[0])
-		buf.Reset()
+	origExpr, err := parsePartitionExpr(ctx, pi.Expr, schema, names)
+	if err != nil {
+		return nil, err
 	}
-	exprs, err := expression.ParseSimpleExprsWithNames(ctx, pi.Expr, schema, names)
+	exprs, err := rewritePartitionExpr(ctx, origExpr, schema, names)
 	if err != nil {
 		// If it got an error here, ddl may hang forever, so this error log is important.
 		logutil.BgLogger().Error("wrong table partition expression", zap.String("expression", pi.Expr), zap.Error(err))
 		return nil, errors.Trace(err)
 	}
-	exprs[0].HashCode(ctx.GetSessionVars().StmtCtx)
-	if col, ok := exprs[0].(*expression.Column); ok {
+	exprs.HashCode(ctx.GetSessionVars().StmtCtx)
+	if col, ok := exprs.(*expression.Column); ok {
 		column = col
 	}
 	return &PartitionExpr{
-		Column: column,
-		Expr:   exprs[0],
-		Ranges: partitionPruneExprs,
+		Column:   column,
+		Expr:     exprs,
+		OrigExpr: origExpr,
 	}, nil
 }
 
+// GetOriginPartitionExpr is used for PointGet plan.
+func (t *partitionedTable) GetOriginPartitionExpr() ast.ExprNode {
```

Comment:
How about this?

```
func getOriginPartitionExpr(pt PartitionedTable) {
    return pt.PartitionExpr().OrigExpr
}
```

Always be careful when adding new methods to an interface.

Reply:
`PartitionExpr()` can not use currently